certify-service/src/main/java/io/mosip/certify/services/KeymanagerLibSigner.java

```diff
@@ -0,0 +1,96 @@
+
+/**
+ * KeymanagerLibSigner is a VCSigner which uses the Certify embedded
+ * keymanager to perform VC signing tasks for JSON LD VCs.
+ * These are the known external requirements:
+ * - the public key must be pre-hosted for the VC & should be available
+ *    so long that VC should be verifiable
+ * - the VC should have a validFrom or issuanceDate in a specific UTC format,
+ *  if missing it uses current time for proof creation timestamp.
+ */
+@Slf4j
+@Service
+public class KeymanagerLibSigner implements VCSigner {
+
+    @Autowired
+    ProofSignatureStrategy signProps;
+    @Value("${mosip.certify.data-provider-plugin.issuer-public-key-uri}")
+    private String hostedKey;
+
+    @Override
+    public VCResult<JsonLDObject> perform(String templatedVC) {
+        // Can the below lines be done at Templating side itself ?
+        VCResult<JsonLDObject> VC = new VCResult<>();
+        JsonLDObject j = JsonLDObject.fromJson(templatedVC);
+        j.setDocumentLoader(null);
+        // NOTE: other aspects can be configured via keyMgrInput map
+        String validFrom;
+        if (j.getJsonObject().containsKey(VCDM1Constants.ISSUANCE_DATE)) {
+            validFrom = j.getJsonObject().get(VCDM1Constants.ISSUANCE_DATE).toString();
+        } else if (j.getJsonObject().containsKey(VCDM2Constants.VALID_FROM)){
+            validFrom = j.getJsonObject().get(VCDM2Constants.VALID_FROM).toString();
+        } else {
+            validFrom = ZonedDateTime.now(ZoneOffset.UTC)
+                    .format(DateTimeFormatter.ofPattern(Constants.UTC_DATETIME_PATTERN));
+        }
+        // TODO: VC Data Model spec doesn't specify a single date format or a
+        //  timezone restriction, this will have to be supported timely.
+        Date createDate = Date
+                .from(LocalDateTime
+                        .parse(validFrom,
+                                DateTimeFormatter.ofPattern(Constants.UTC_DATETIME_PATTERN))
+                        .atZone(ZoneId.systemDefault()).toInstant());
+        LdProof vcLdProof = LdProof.builder().defaultContexts(false).defaultTypes(false).type(signProps.getName())
+                .created(createDate).proofPurpose(VCDMConstants.ASSERTION_METHOD)
+                .verificationMethod(URI.create(hostedKey))
+                .build();
+        // 1. Canonicalize
+        Canonicalizer canonicalizer = signProps.getCanonicalizer();
+        byte[] vcSignBytes = null;
+        try {
+            vcSignBytes = canonicalizer.canonicalize(vcLdProof, j);
+        } catch (IOException | GeneralSecurityException | JsonLDException e) {
+            log.error("Error during canonicalization", e.getMessage());
+            throw new CertifyException("Error during canonicalization");
+        }
+        String vcEncodedHash = Base64.getUrlEncoder().encodeToString(vcSignBytes);
+        String sign = signProps.getProof(vcEncodedHash);
+        LdProof ldProofWithJWS = signProps.buildProof(vcLdProof, sign);
```

Comment:
above 2 method names of signProps are confusing, can we merge both and keep it as a single call like generateLDProof passinging the vcEncodedHash and get the LDProof ?

Also ldProofWithJWS to be renamed since the JWS usage is not always true in all ProofSignatureStrategy

Reply:
This together with the expectation that the applicationId & referenceId should be passed from CertifyIssuanceServiceImpl makes this interface slightly more tied down to Keymanager's way of identifying a keypair with two String identifiers. Is this still a good idea? With this, we might make Certify indirectly more dependent on Keymanager(MOSIP Platform component).

@vishwa-vyom wdyt?

Reply:
Lets pass the uri to generateLDProof and avoid the LdProod object passing